### PR TITLE
Add Agents Catalog Foundation

### PR DESCRIPTION
**Pull request by Philip-Carneiro**
## Description

Implemented the complete data foundation layer for the Agents Catalog frontend. 
Followed MCP Catalog pattern.

- Non-PF
Screenshot 2026-07-08 at 10 50 49

- PF
Screenshot 2026-07-08 at 10 51 07

## Merge criteria:

- All the commits have been [_signed-off_](https://github.com/kubeflow/community/tree/master/dco-signoff-hook#signing-off-commits)  (To pass the `DCO` check)

- [x] The commits have meaningful messages
- [x] Automated tests are provided as part of the PR for major new functionalities; testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work.
- [x] Code changes follow the [kubeflow contribution guidelines](https://www.kubeflow.org/docs/about/contributing/).
- [ ] **For first time contributors**: Please reach out to the [Reviewers](https://github.com/kubeflow/hub/blob/main/OWNERS) to ensure all tests are being run, ensuring the label `ok-to-test` has been added to the PR.

If you have UI changes

- [x] The developer has added tests or explained why testing cannot be added.
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Verify that UI/UX changes conform the UX guidelines for Kubeflow.

### CODE REVIEW

File: clients/ui/frontend/src/app/hooks/agentsCatalog/useAgentFilterOptionList.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  }
+  const mapped: AgentsCatalogFilterOptionsList['filters'] = {};
+  for (const [key, value] of Object.entries(raw.filters)) {
+    const frontendKey = BACKEND_TO_FRONTEND_AGENT_FILTER_KEY[key] ?? key;
+    if (isAgentFilterCategoryKey(frontendKey) && isFilterStringOption(value)) {
+      mapped[frontendKey] = value;
+    }
+  }
+  return { filters: mapped };
+}
```
Comment (manaswinidas) [inline clients/ui/frontend/src/app/hooks/agentsCatalog/useAgentFilterOptionList.ts:41]: `mapBackendFilterOptions` is the only function in this PR with real branching logic (key remapping, type guards via `isFilterStringOption` using `Object.getOwnPropertyDescriptor`), yet it has zero test coverage.

Please add a unit test covering at minimum:
- Backend key remapped correctly (`tags` → `labels`)
- Unknown keys are dropped
- Non-string filter options are dropped
- `null`/`undefined` filters input returns `{ filters: undefined }`

This is where silent failures are most likely — a bug here means the filter panel renders empty with no visible error.

File: clients/ui/frontend/src/app/hooks/agentsCatalog/useAgentsBySourceLabel.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    setAllItems([]);
+    setTotalSize(0);
+    setNextPageTokenValue('');
+    setIsLoadingMore(false);
+    setLoadMoreError(undefined);
+    refetch();
+  }, [refetch]);
+
+  const paginatedData: PaginatedAgentList = {
+    items: allItems,
```
Comment (manaswinidas) [inline clients/ui/frontend/src/app/hooks/agentsCatalog/useAgentsBySourceLabel.ts]: This object literal is recreated on every render without `useMemo`, producing a new reference each time. Downstream consumers (e.g., `CatalogGalleryLayout`) receiving this as a prop will re-render unnecessarily.

Wrap in `useMemo` with appropriate deps, or restructure to return stable references.

File: clients/ui/frontend/src/app/hooks/agentsCatalog/useAgentsBySourceLabel.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      setLoadMoreError(undefined);
+    } catch (err) {
+      setLoadMoreError(
+        new Error(
+          `Failed to load more agents: ${err instanceof Error ? err.message : String(err)}`,
+        ),
+      );
+    } finally {
+      setIsLoadingMore(false);
+    }
```
Comment (manaswinidas) [inline clients/ui/frontend/src/app/hooks/agentsCatalog/useAgentsBySourceLabel.ts:123]: Including `isLoadingMore` in the dependency array means `loadMore`'s identity changes on every loading state toggle. When passed as a prop to gallery components, this triggers child re-renders.

Consider using a ref (`useRef`) for the guard check instead:
```ts
const isLoadingMoreRef = React.useRef(false);
```

File: clients/ui/frontend/src/app/agentsCatalogTypes.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  env?: AgentEnvVar[];
+  artifacts?: AgentArtifact[];
+  customProperties?: Record<string, unknown>;
+  createTimeSinceEpoch?: string;
+  lastUpdateTimeSinceEpoch?: string;
+};
+
+export type AgentList = PaginationParams & { items?: Agent[] };
+
+export type AgentListParams = {
```
Comment (manaswinidas) [inline clients/ui/frontend/src/app/agentsCatalogTypes.ts:38]: nit: `pageSize` is typed as `number | string` but the service layer always does `String(listParams.pageSize)` internally. This leaks a serialization concern into the domain type. Consider typing as just `number`.

File: clients/ui/frontend/src/__mocks__/mockAgentsCatalogTestData.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+export const mockAgentList = (partial?: Partial<AgentList>): AgentList => ({
+  items: [mockAgent()],
+  pageSize: 10,
+  size: 1,
+  nextPageToken: '',
+  ...partial,
+});
+
+export const mockAgents: Agent[] = [
```
Comment (manaswinidas) [inline clients/ui/frontend/src/__mocks__/mockAgentsCatalogTestData.ts]: nit: `mockAgents` is exported but never imported — not in this PR, and not in #2934 (which introduces `mockAgentsGallery` instead). Consider removing to avoid dead code, or aligning with the follow-up PR's naming.

File: clients/ui/frontend/src/app/pages/agentsCatalog/const.ts [hunk 6]
```diff
@@ -1,3 +1,16 @@
...
   'Discover agents that are available for your organization.';
+
+export const AGENT_FILTER_KEYS: AgentFilterCategoryKey[] = ['framework', 'labels'];
+
+export const AGENT_FILTER_CATEGORY_NAMES: Record<AgentFilterCategoryKey, string> = {
+  framework: 'Framework',
+  labels: 'Labels',
+};
+
+export const BACKEND_TO_FRONTEND_AGENT_FILTER_KEY: Record<string, AgentFilterCategoryKey> = {
```
Comment (manaswinidas) [inline clients/ui/frontend/src/app/pages/agentsCatalog/const.ts]: This PR defines `AgentFilterCategoryKey = 'framework' | 'labels'` with a `tags → labels` mapping, but #2934 immediately replaces this with `'framework' | 'category' | 'communicationProtocol' | 'testedModels'` and empties the mapping. Consider aligning with the final schema now to avoid throwaway code.

**Review by manaswinidas — CHANGES_REQUESTED** [review]
Good foundation work — follows the MCP catalog pattern consistently. Requesting changes for one blocking item:

**Blocking:** `mapBackendFilterOptions` needs a unit test. It's the only function with real branching logic and no coverage — a bug here silently breaks the filter panel with no visible error.

**Non-blocking suggestions** (can address in follow-up):
- `paginatedData` object should be memoized to avoid unnecessary re-renders
- `isLoadingMore` in `loadMore` deps causes identity churn — use a ref
- Align filter keys with #2934's final schema to avoid throwaway code
- Remove unused `mockAgents` export
- Type `pageSize` as `number` only

File: clients/ui/frontend/src/app/context/agentsCatalog/__tests__/AgentsCatalogContext.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+  it('updates filters via setFilters', () => {
+    const { result } = renderHook(() => React.useContext(AgentsCatalogContext), { wrapper });
+    act(() => {
+      result.current.setFilters({ framework: ['LangGraph'] });
+    });
+    expect(result.current.filters).toEqual({ framework: ['LangGraph'] });
+    act(() => {
+      result.current.setFilters((prev) => ({ ...prev, labels: ['research'] }));
+    });
```
Comment (manaswinidas) [inline clients/ui/frontend/src/app/context/agentsCatalog/__tests__/AgentsCatalogContext.spec.tsx:99]: nit: `labels` is no longer a valid `AgentFilterCategoryKey` after the realignment — valid keys are now `framework | category | communicationProtocol | testedModels`. This test passes at runtime but uses an invalid key.

Please update to use a valid key (e.g., `category: ['General purpose']`) in #2934 when you touch this file.

**Review by manaswinidas — APPROVED** [review]
All review comments addressed — nice work. One minor nit left (stale `labels` key in the context spec) that can be fixed in #2934.

**Review by manaswinidas — APPROVED** [review]
/approve
